ext/rugged/rugged_rebase.c

```diff
@@ -0,0 +1,263 @@
+	if (RTEST(val))
+		ret->quiet = 1;
+
+	val = rb_hash_aref(rb_options, CSTR2SYM("rewrite_notes_ref"));
+	if (!NIL_P(val)) {
+		Check_Type(val, T_STRING);
+		ret->rewrite_notes_ref = StringValueCStr(val);
+	}
+
+	rugged_parse_checkout_options(&ret->checkout_options, rb_options);
+}
+
+void rb_git_rebase__free(git_rebase *rebase)
+{
+	git_rebase_free(rebase);
+}
+
+VALUE rugged_rebase_new(VALUE klass, VALUE owner, git_rebase *rebase)
+{
+	VALUE rb_rebase = Data_Wrap_Struct(klass, NULL, &rb_git_rebase__free, rebase);
+	rugged_set_owner(rb_rebase, owner);
+	return rb_rebase;
+}
+
+static VALUE rb_git_rebase_new(int argc, VALUE* argv, VALUE klass)
+{
+	int error;
+	git_rebase *rebase;
+	git_repository *repo;
+	git_annotated_commit *branch = NULL, *upstream = NULL, *onto = NULL;
+	git_oid id;
+	VALUE rb_repo, rb_branch, rb_upstream, rb_onto, rb_options;
+	git_rebase_options options = GIT_REBASE_OPTIONS_INIT;
+
+	rb_scan_args(argc, argv, "31:", &rb_repo, &rb_branch, &rb_upstream, &rb_onto, &rb_options);
+	Data_Get_Struct(rb_repo, git_repository, repo);
+	Check_Type(rb_branch, T_STRING);
+	Check_Type(rb_upstream, T_STRING);
+
+	if ((error = git_oid_fromstr(&id, StringValueCStr(rb_branch))) < 0 ||
+	    (error = git_annotated_commit_lookup(&branch, repo, &id)) < 0 ||
+	    (error = git_oid_fromstr(&id, StringValueCStr(rb_upstream))) < 0 ||
+	    (error = git_annotated_commit_lookup(&upstream, repo, &id)) < 0)
+		goto cleanup;
+
+	if (!NIL_P(rb_onto)) {
+		Check_Type(rb_onto, T_STRING);
+		if ((error = git_oid_fromstr(&id, StringValueCStr(rb_onto))) < 0 ||
+		    (error = git_annotated_commit_lookup(&onto, repo, &id)) < 0)
+			goto cleanup;
+	}
+
+	parse_rebase_options(&options, rb_options);
+	error = git_rebase_init(&rebase, repo, branch, upstream, onto, &options);
+
+cleanup:
+	git_annotated_commit_free(branch);
+	git_annotated_commit_free(upstream);
+	git_annotated_commit_free(onto);
+
+	rugged_exception_check(error);
+
+	return rugged_rebase_new(klass, rb_repo, rebase);
+}
+
+static VALUE rb_git_rebase_next(VALUE self)
+{
+	int error;
+	git_rebase *rebase;
+	git_rebase_operation *operation;
+
+	Data_Get_Struct(self, git_rebase, rebase);
+	error = git_rebase_next(&operation, rebase);
+	if (error == GIT_ITEROVER)
+		return Qnil;
+
+	rugged_exception_check(error);
+
+	return ruged_rebase_operation_new(operation, self);
+}
+
+static VALUE rb_git_rebase_commit(VALUE self, VALUE rb_author, VALUE rb_committer, VALUE rb_message)
+{
+	int error;
+	git_oid id;
+	git_rebase *rebase;
+	git_signature *author = NULL, *committer;
+	const char *message = NULL;
+
+	Data_Get_Struct(self, git_rebase, rebase);
+	if (!NIL_P(rb_message)) {
+		Check_Type(rb_message, T_STRING);
+		message = StringValueCStr(rb_message);
+	}
+
+	if (NIL_P(rb_committer))
+		rb_raise(rb_eArgError, "Expected non-nil committer");
+	else
+		committer = rugged_signature_get(rb_committer, NULL);
+
+	if (!NIL_P(rb_author))
+		author = rugged_signature_get(rb_author, NULL);
```

Comment:
Yes, I've been thinking about this today after my first comment. If you want to catch a Ruby Exception in C-Land, you need to wrap the call using `VALUE rb_protect(VALUE (*)(VALUE), VALUE, int*);`.

We could add wrapper functions for things like `rugged_signature_get` that take all arguments, stash them into a struct, pass this down `rb_protect` with another function that unwraps that struct and calls the original `rugged_signature_get` again.

But that feels really convoluted and wasteful. You might apply some macro magic to make this look nicer, but the wrapping/unwrapping would still be there.

Another approach would be to split functions like `rugged_signature_get` into 2 separate functions like `rugged_signature_check` (which performs all the checks) and `rugged_signature_get` (which performs none of the checks and only handles conversion).

I think I favor the latter approach.